core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1217,26 +1217,41 @@ public String toString() {
      * @return true if the remote segment's epoch/offsets are within the leader epoch lineage of the partition.
      */
     // Visible for testing
-    public static boolean isRemoteSegmentWithinLeaderEpochs(RemoteLogSegmentMetadata segmentMetadata,
-                                                            long logEndOffset,
-                                                            NavigableMap<Integer, Long> leaderEpochs) {
+    static boolean isRemoteSegmentWithinLeaderEpochs(RemoteLogSegmentMetadata segmentMetadata,
+                                                     long logEndOffset,
+                                                     NavigableMap<Integer, Long> leaderEpochs) {
         long segmentEndOffset = segmentMetadata.endOffset();
         // Filter epochs that does not have any messages/records associated with them.
         NavigableMap<Integer, Long> segmentLeaderEpochs = buildFilteredLeaderEpochMap(segmentMetadata.segmentLeaderEpochs());
         // Check for out of bound epochs between segment epochs and current leader epochs.
-        Integer segmentFirstEpoch = segmentLeaderEpochs.firstKey();
         Integer segmentLastEpoch = segmentLeaderEpochs.lastKey();
-        if (segmentFirstEpoch < leaderEpochs.firstKey() || segmentLastEpoch > leaderEpochs.lastKey()) {
+        if (segmentLastEpoch < leaderEpochs.firstKey() || segmentLastEpoch > leaderEpochs.lastKey()) {
```

Comment:
I'm not sure if the check `segmentLastEpoch < leaderEpochs.firstKey()` makes sense or not.

Suppose: 
leader-epoch-file-cache: {(5, 10), (7, 15), (9, 100)}
segment1: offset-range = 5-50, Broker = 0, epochs = {(5, 10), (7, 15)}
Now, delete_records are called and log start offset incremented to 100, so the new leader-epoch-file-cache will be: {(9, 100)}

When entering this check, it'll fail because the segmentLastEpoch (7) will be < leaderEpochs.firstKey() (9). But we still want to delete this segment, right?

Reply:
For this case, the segment will be deleted by the `isSegmentBreachByLogStartOffset` check and it won't enter `isRemoteSegmentWithinLeaderEpochs` check.

Reply:
I'll add integration tests with LocalTieredStorage to cover various scenarios. The test will be useful to catch any regression if we refactor the code later.